src/main/java/at/bitfire/vcard4android/ContactReader.kt

```diff
@@ -42,12 +45,10 @@ class ContactReader internal constructor(val vCard: VCard, val downloader: Conta
             val date = prop.date
             if (prop.partialDate == null && date != null) {
                 prop.getParameter(Contact.DATE_PARAMETER_OMIT_YEAR)?.let { omitYearStr ->
-                    val cal = GregorianCalendar.getInstance()
-                    cal.time = date
-                    if (cal.get(GregorianCalendar.YEAR).toString() == omitYearStr) {
+                    if (date.get(ChronoField.YEAR).toString() == omitYearStr) {
                         val partial = PartialDate.builder()
-                            .date(cal.get(GregorianCalendar.DAY_OF_MONTH))
-                            .month(cal.get(GregorianCalendar.MONTH) + 1)
+                            .date(date.get(ChronoField.DAY_OF_MONTH))
+                            .month(date.get(ChronoField.MONTH_OF_YEAR))
```

Comment:
Is it guaranteed that prop.date has these ChronoFields? Otherwise, we can't rely on it.

For instance, the ez-vcard documentation says the date could be an `Instant`. I think `Instant` doesn't support DAY_OF_MONTH, YEAR and MONTH_OF_YEAR: https://developer.android.com/reference/java/time/Instant?hl=en#isSupported(java.time.temporal.TemporalField)

Reply:
Yes, this is in fact an issue. I've tried this test inside `ContactReaderTest`:
```kotlin
@Test
fun testCheckPartialDate_Temporal() {
    val prop = DateOrTimeProperty(Instant.ofEpochSecond(1680787224)).apply {
        addParameter(Contact.DATE_PARAMETER_OMIT_YEAR, "2023")
    }
    ContactReader.checkPartialDate(prop)
}
```
and it fails with
```
Unsupported field: Year
java.time.temporal.UnsupportedTemporalTypeException: Unsupported field: Year
	at java.base/java.time.Instant.get(Instant.java:565)
	at at.bitfire.vcard4android.ContactReader$Companion.checkPartialDate(ContactReader.kt:48)
	at at.bitfire.vcard4android.ContactReaderTest.testCheckPartialDate_Temporal(ContactReaderTest.kt:584)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:108)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:57)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:39)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:52)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```
Since with `Instant` we really can only relay on seconds and millis. So I think that maybe the best option is that if we need to extract that year field, we convert from `Instant` to `LocalDateTime`, which can directly be done with `LocalDateTime.ofInstant`, for example, like this:
```kotlin
fun checkPartialDate(prop: DateOrTimeProperty) {
    val date = prop.date
    if (prop.partialDate == null && date != null) {
        prop.getParameter(Contact.DATE_PARAMETER_OMIT_YEAR)?.let { omitYearStr ->
            // If date is instant, convert to LocalDateTime so year can be extracted
            val convertedDate = date.takeIf { it !is Instant }
                ?: LocalDateTime.ofInstant(date as Instant, TimeZone.getDefault().toZoneId())

            if (convertedDate.get(ChronoField.YEAR).toString() == omitYearStr) {
                val partial = PartialDate.builder()
                    .date(convertedDate.get(ChronoField.DAY_OF_MONTH))
                    .month(convertedDate.get(ChronoField.MONTH_OF_YEAR))
                    .build()
                prop.partialDate = partial
            }

            // X-APPLE-OMIT-YEAR not required anymore because we're now working with PartialDate
            prop.removeParameter(Contact.DATE_PARAMETER_OMIT_YEAR)
        }
    }
}
```
Maybe it's a bit dirty, but I think it's the easiest, quickest and safest approach. Btw this modification does pass the test.

Reply:
I have looked through the ez-vcard 0.12 code and they seem to use `Instant` only for `REV` (`Revsion`).

Our `checkPartialDate` is however used for `BDAY`, `ANNIVERSARY` and `X-AB-DATE`, which are expressed as `LocalDate`, if I understand it correctly.

So my suggestion is:

- please double-check that `Instant` is currently only used for `REV`  → so `checkPartialDate` won't be fed with `Instant`
- check the other types (`LocalDate`, `LocalDateTime`, `OffsetDateTime`) and whether they're interesting for `checkPartialDate` or not
- document the types which `checkPartialDate` shall handle in its KDoc
- handle only these types and maybe generate a warning and ignore if the type is something else

Reply:
We can also do
```kotlin
// Check that all the required fields are supported.
if (!listOf(ChronoField.YEAR, ChronoField.MONTH_OF_YEAR, ChronoField.DAY_OF_MONTH).all { date.isSupported(it) }){
    Constants.log.warning("Passed a Temporal into checkPartialDate that doesn't support the YEAR field.")
    return
}
```
But I think it might be too much, and a bit verbose. What do you also think @sunkup ?

Reply:
We need `YEAR`, `MONTH_OF_YEAR` and `DAY_OF_MONTH` so I'd check for all three.

I however suggest to

- invert the logic expression: `arrayOf(…).any { !date.isSupported(it) }` – if any of these fields is not supported, print warning and ignore it
- using `arrayOf()` instead of `listOf()` is probably more efficient, because an Array object is simpler than a List object